### DISCUSSION TITLE
How to use webtorrent in a Svelte 5 application?

### DISCUSSION THREAD

**Discussion by rgl** [webtorrent/webtorrent#2887] [Q&A]
Can you please show how to adapt the https://webtorrent.io/intro example to run in svelte 5 (using sveltekit) application?

I've tried it in a new svelte `routes/film/+page.svelte` file as:

```html
<script lang="ts">
    import { onMount } from 'svelte';
    import WebTorrent from 'webtorrent'; // <-- it fails here.

    let webtorrentElement: HTMLElement;
</script>

<div bind:this={webtorrentElement}>
    TODO
</div>
```

But when that page is loaded in the browser, the following error appears in the browser console:

```
Module "events" has been externalized for browser compatibility. Cannot access "events.EventEmitter" in client code. See https://vite.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
```

Do you have any idea how to fix this? Can you point me in the right direction?

**Comment by osenvosem** [comment]
Try importing pre-build version like "import WebTorrent from 'webtorrent/dist/webtorrent.min.js'" or make alias in your builder.

**Reply by SilentBot1** [reply]
This should be the go-to way for now, the WebTorrent build process isn't the greatest. Additionally, you will need to only CSR WebTorrent, it can not be server-side rendered.

**Reply by osenvosem** [reply]
The build process is downright bad as well as the project governance. Still didn't manage to make it working on the frontend.

**Comment by dukeh3** [comment]
I have managed to make it work in my demo app. what you need to do is the following.

1. import the min.js files, so

    import SimplePeer from "simple-peer/simplepeer.min"
    import WebTorrent from 'webtorrent/dist/webtorrent.min'

2. disable SSR

so put the following into your +page.ts (or +page.js)

export const ssr = false;

3. When loading the sw you need to copy it from node_modules and put it in the static root during the build process.

3.1 install the copy plugin

npm install vite-plugin-static-copy --save-dev

3.2 update vite.config.ts

export default defineConfig({
	plugins: [
		sveltekit(),
		viteStaticCopy({
			targets: [
				{
					src: 'node_modules/webtorrent/dist/sw.min.js', // Path to the service worker in node_modules
					dest: '' // Copies to the root of the `static` directory
				}
			]
		})
	]
});

3.3 when loading the sw worker make sure the path is correct (./ as written in the examples on the net is NOT the correct path), you can set scope to whatever you want

navigator.serviceWorker.register('/sw.min.js', {scope: './'}).then(reg => {
